src/main/java/org/varroa/reporting/AgentAggregator.java

```diff
@@ -56,10 +65,45 @@ public void run() {
                     actionReports.computeIfAbsent(queueEntry.actionId, actionId -> new ActionReport());
             actionReport.addDuration(queueEntry.duration);
 
+            if (actionReports.size() > 50) {
+                sendReportingMessage();
+            }
+
             drained++;
         }
     }
 
+    public @NotNull CompletableFuture<Void> clearQueueAndSendReportingMessage() {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        return clearQueueAndSendReportingMessage(future);
+    }
+
+    private @NotNull CompletableFuture<Void> clearQueueAndSendReportingMessage(
+            final @NotNull CompletableFuture<Void> future) {
+        executorService.submit(() -> {
+            if (queueCounter.get() == 0) {
+                sendReportingMessage();
+                future.complete(null);
+            } else {
+                executorService.submit(() -> clearQueueAndSendReportingMessage(future));
```

Comment:
```suggestion
                clearQueueAndSendReportingMessage(future);
```
The method does already schedule